public/app/features/alerting/unified/components/receivers/AlertInstanceModalSelector.tsx

```diff
@@ -0,0 +1,381 @@
+import { css, cx } from '@emotion/css';
+import React, { CSSProperties, useCallback, useEffect, useMemo, useState } from 'react';
+import AutoSizer from 'react-virtualized-auto-sizer';
+import { FixedSizeList } from 'react-window';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import {
+  Button,
+  clearButtonStyles,
+  FilterInput,
+  LoadingPlaceholder,
+  Modal,
+  Tooltip,
+  useStyles2,
+  Icon,
+  Tag,
+} from '@grafana/ui';
+import { AlertmanagerAlert, TestTemplateAlert } from 'app/plugins/datasource/alertmanager/types';
+import { dispatch } from 'app/store/store';
+
+import { useUnifiedAlertingSelector } from '../../hooks/useUnifiedAlertingSelector';
+import { fetchAmAlertsAction } from '../../state/actions';
+import { GRAFANA_RULES_SOURCE_NAME } from '../../utils/datasource';
+import { arrayLabelsToObject, labelsToTags, objectLabelsToArray } from '../../utils/labels';
+import { initialAsyncRequestState } from '../../utils/redux';
+import { extractCommonLabels, omitLabels } from '../rules/state-history/common';
+
+export function AlertInstanceModalSelector({
+  onSelect,
+  isOpen,
+  onClose,
+}: {
+  onSelect: (alerts: TestTemplateAlert[]) => void;
+  isOpen: boolean;
+  onClose: () => void;
+}) {
+  const styles = useStyles2(getStyles);
+
+  const [selectedRule, setSelectedRule] = useState<string>();
+  const [selectedInstances, setSelectedInstances] = useState<AlertmanagerAlert[] | null>(null);
+
+  const [filteredRules, setFilteredRules] = useState<Record<string, AlertmanagerAlert[]> | null>(null);
+
+  const alertsRequests = useUnifiedAlertingSelector((state) => state.amAlerts);
+  const { loading, result, error } = alertsRequests[GRAFANA_RULES_SOURCE_NAME] || initialAsyncRequestState;
+
+  const [ruleFilter, setRuleFilter] = useState('');
+
+  useEffect(() => {
+    dispatch(fetchAmAlertsAction(GRAFANA_RULES_SOURCE_NAME));
+  }, []);
+
+  const rulesWithInstances: Record<string, AlertmanagerAlert[]> = useMemo(() => {
+    return {};
+  }, []);
+
+  const handleRuleChange = useCallback((rule: string) => {
+    setSelectedRule(rule);
+    setSelectedInstances(null);
+  }, []);
+
+  useEffect(() => {
+    if (!loading && result) {
+      result.forEach((instance) => {
+        if (!rulesWithInstances[instance.labels['alertname']]) {
+          rulesWithInstances[instance.labels['alertname']] = [];
+        }
+        rulesWithInstances[instance.labels['alertname']].push(instance);
+      });
+      setFilteredRules(rulesWithInstances);
+    }
+  }, [loading, result, rulesWithInstances]);
+
+  if (error) {
+    return null;
+  }
+
+  const RuleRow = ({ index, style }: { index: number; style?: CSSProperties }) => {
+    if (!filteredRules) {
+      return null;
+    }
+    const ruleName = Object.keys(filteredRules)[index];
```

Comment:
It might be worth creating a separate variable for `Object.keys(filteredRules)` so we don't invoke `Object.keys` for each `RuleRow`